### DISCUSSION TITLE
What is the AdcParam for MQ135 analog sensor?

### DISCUSSION THREAD

**Discussion by valdus2** [arendst/Tasmota#20051] [Q&A]
I can see in tasmota docs for example that for mq131 is "AdcParam 10, 131.00, 23.94, -1.11, 15.00" but not for mq135.
I don't know where to find the ANALOG_MQ_A, ANALOG_MQ_B and ANALOG_MQ_RatioMQCleanAir (params 2.3.4).
Thank you in advance.

**Comment by valdus2** [comment]
Please, anyone any help, any guidance ...?

**Comment by joba-1** [comment]
Can't help with a set of values, just with my experience with those sensors:

I have 2 of them active for a long time with my own firmware on arduino nanos, but the insight I get is very limited:

* both show very different values in the same spot and the same hardware: each has to be calibrated separately.
* both drift a lot: values in fresh air vary from day to day (but not all over the place, like +-10%): calibrate each day.
* they do not only react on CO2: also other chemicals/gases, temperature, air convection/wind and probably more have an influence: only valid CO2 data in places where the environment does not change, except for the CO2.

Calibration works by placing them in clean air for 10min and then declare the adc value as representing the current global average CO2 concentration (last year: 421ppm, when I started it was close to 400ppm!). 
If the sensors are in a room where you know you have fresh air from time to time, you can calibrate by using the lowest measurement as representing clean air (you only get lower ppm than in clean air in unusual environments...)

If you are still interested anyways: maybe I can find my firmware source code to look up the formula I used to calculate how change in adc values influence change in ppm... :)

**Comment by valdus2** [comment]
Thanks for the info. I have already 5 of them and the CO2 and smoke is what I care are about mostly.
They're all working well as of now, reading them with 5V from a Denkovi smartDEN Notifier, showing same concentration in one room

I just don't know how to implement them with analog input in tasmota as per https://tasmota.github.io/docs/ADC/#commands:

    <param1> = ANALOG_MQ_TYPE (default 2) It used to specify sensor type. At the moment exists: 2, 3, 4, 5, 6, 7, 8, 9, 131, 135 (means MQ-02, MQ-03, MQ-04 ecc.).
    <param2> = ANALOG_MQ_A (default 574.25 a params for MQ-02) It is exponential regression a params
    <param3> = ANALOG_MQ_B (default -2.222 b params for MQ-02) It is exponential regression b params, generally negative
    <param4> = ANALOG_MQ_RatioMQCleanAir (default 15.0 RatioMQCleanAir params for MQ-02) NOT USED YET. It is threshold for good air in ppm for future alams arming

    Usage example for MQ-02, MQ-04, MQ-07 and MQ-131
    AdcParam 10, 2.00, 574.25, -2.22, 9.83
    AdcParam 10, 4.00, 1012.70, -2.79, 4.40
    AdcParam 10, 7.00, 99.04, -1.52, 27.50
    AdcParam 10, 131.00, 23.94, -1.11, 15.00


So, for mq135 that would start with 

    AdcParam 10, 135.00, param2, param3, param4

And the question is what are the param2, 3, 4 for the MQ135?

**Reply by sfromis** [reply]
This driver seems to have config settings for various usages of that thing (which has no capabilities for measuring anything specific)
https://github.com/miguel5612/MQSensorsLib/blob/master/examples/MQ-135-ALL/MQ-135-ALL.ino
Thus, there can be no single "right" set of parameters, and any selection can't work well if other things in the air are also variable.

It seems to be just another example of using tin dioxide, which reacts to a lot of different things in the air, and therefore can't actually measure anything. What it can do is to register that something is different to a degree (higher/lower resistance), and if you assume that only one factor is likely to change, you can, sort of, get an indication of how much. Takes specific calibration.

In a non-industrial setting, it is not a thing I'd like to use for numbers, but potentially for "better look at what is happening". Makes more sense in an industrial setting, like monitoring expected values of ammonia, benzene and other "problematic" gasses.

**Reply by valdus2** [reply]
Thank you for the info. As you said : 

> What it can do is to register that something is different to a degree (higher/lower resistance), and if you assume that only one factor is likely to change, you can, sort of, get an indication of how much.

This is how I use it because I have already calibrate them and I know the range between clean and not clean air.
The question has to deal with the initial installation of MQ135 with Tasmota.
Do I have to set the AdcParam or am I OK by just configuring an analog input for mq sensor and just take the measured value ?

**Reply by sfromis** [reply]
Well, the driver I linked does have what looks like such settings for the MQ135, which probably make sense as a starting point.

**Comment by valdus2** [comment]
Sorry for the late answer bur that was a direct hit from corona virus, I'm OK now.

For a start I want to thank you for the info and leading me in this driver. 

So, I'm almost there, if the below assumption is correct:

In the driver that you led me, for every measured concentration value there are two values A and B.
If these two values (A and B) correspond to ANALOG_MQ_A and ANALOG_MQ_B in Tasmota, 
then what should be the 4th parameter ANALOG_MQ_RatioMQCleanAir ?

**Reply by sfromis** [reply]
IDK, try just some number..... The docs **which you already quoted** says:
> NOT USED YET. It is threshold for good air in ppm for future alams arming

**Reply by valdus2** [reply]
OK, I'm gonna try it that way like you said sfromis. Thank you again for your help.

**Reply by photomax2202** [reply]
Hi valdus2,

did you get valid values from mq135 with Tasmota.

I used param values from linked Arduino-File for test. But I did not get valid values.
My param command.

AdcParam 10, 135.00, 110.47, -2.862, 3.60

CleanAirRatio from Calibration.ino

**Reply by photomax2202** [reply]
May someone did have more math power than myself:

There is an German Blog Article witch is explaining an MQ-2 Sample.
https://tutorials-raspberrypi.de/raspberry-pi-gas-sensor-mq2-konfigurieren-und-auslesen/?utm_content=cmp-true

With Datasheet of MQ-135 I can build log10-Function
https://www.wolframalpha.com/input?i=%28+Log10%280.8%29+-++log10%282.3%29+%29+%2F+%28+log10%28200%29+-+log10%2810%29+%29

 But my math power is not high enough to calculate from pitch the x and y point from straight line. 

Ok ive tried:
y=-0.352519 x + 0.714247

But actually no idea for exponential regression a & b

 May someone can help???

**Reply by valdus2** [reply]
The last value 3.60, where did it came from? Shouldn't be 0 as per sfromis?

**Reply by sfromis** [reply]
It just does not matter. Use whatever value, and 0 may be "less confusing".....

**Reply by photomax2202** [reply]
https://github.com/miguel5612/MQSensorsLib/blob/master/examples/Calibration/Calibration.ino

**Reply by photomax2202** [reply]
May I just made a little mistake.

I do have 5 boards of:
https://www.google.com/search?q=az-envy

And I changed Sensors (MQ2 to MQ135). But I missed changing the divider resistors.

Edit:
There is just another fault. My dev-borard does have an internal ADC divider to get add range 0...3.3 V. This is a little bit tricky.

**Comment by MaxDaNr1** [comment]
Hi there,
I have quite the same problem with my MQ-5 sensor on Tasmota. I have no idea with parameters i have to set with
`ADCparam 10 5.0, param2, param3, param4`
I have no idea which parameters to use. Can anybody help me?
When I let param2, param3 & param4 on default values I only get 0.63 ppm as value

**Reply by sfromis** [reply]
Do read the suggestions in this thread.....

**Reply by MaxDaNr1** [reply]
I did, but I still have no clue what to do. That's the reason why I am asking.

**Reply by sfromis** [reply]
Read again. There is no concept of a "correct" value, the whole "sensor" cannot measure anything specific (as already explained), only get the resistance of a bit of corroded tin. And I specifically provided a link having several different suggestions of values you could try, and someone else also posted an example.

However, do not expect a clean explanation of exactly what to do, that's not how the "sensor" works. You can try some plausible values, and after that calibrate for your situation. And if you want an "easy" sensor you can just connect and get a reasonably good number, you did not make a good choice.
